pkg/addonmanager/controllers/agentdeploy/healthcheck_sync.go

```diff
@@ -0,0 +1,177 @@
+package agentdeploy
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	addonapiv1alpha1 "open-cluster-management.io/api/addon/v1alpha1"
+	clusterv1 "open-cluster-management.io/api/cluster/v1"
+	workapiv1 "open-cluster-management.io/api/work/v1"
+
+	"open-cluster-management.io/addon-framework/pkg/addonmanager/constants"
+	"open-cluster-management.io/addon-framework/pkg/agent"
+	"open-cluster-management.io/addon-framework/pkg/basecontroller/factory"
+)
+
+type healthCheckSyncer struct {
+	getWorkByAddon func(addonName, addonNamespace string) ([]*workapiv1.ManifestWork, error)
+	agentAddon     agent.AgentAddon
+}
+
+func (s *healthCheckSyncer) sync(ctx context.Context,
+	syncCtx factory.SyncContext,
+	cluster *clusterv1.ManagedCluster,
+	addon *addonapiv1alpha1.ManagedClusterAddOn) (*addonapiv1alpha1.ManagedClusterAddOn, error) {
+	// reconcile health check mode
+	var expectedHealthCheckMode addonapiv1alpha1.HealthCheckMode
+
+	if s.agentAddon.GetAgentAddonOptions().HealthProber == nil {
+		return addon, nil
+	}
+
+	switch s.agentAddon.GetAgentAddonOptions().HealthProber.Type {
+	case agent.HealthProberTypeWork, agent.HealthProberTypeNone:
+		expectedHealthCheckMode = addonapiv1alpha1.HealthCheckModeCustomized
+	case agent.HealthProberTypeLease:
+		expectedHealthCheckMode = addonapiv1alpha1.HealthCheckModeLease
+	default:
+		expectedHealthCheckMode = addonapiv1alpha1.HealthCheckModeLease
+	}
+
+	if expectedHealthCheckMode != addon.Status.HealthCheck.Mode {
+		addon.Status.HealthCheck.Mode = expectedHealthCheckMode
+	}
+
+	err := s.probeAddonStatus(addon)
+	return addon, err
+}
+
+func (s *healthCheckSyncer) probeAddonStatus(addon *addonapiv1alpha1.ManagedClusterAddOn) error {
+	if s.agentAddon.GetAgentAddonOptions().HealthProber.Type != agent.HealthProberTypeWork {
+		return nil
+	}
+
+	if s.agentAddon.GetAgentAddonOptions().HealthProber.WorkProber == nil {
+		meta.SetStatusCondition(&addon.Status.Conditions, metav1.Condition{
+			Type:    addonapiv1alpha1.ManagedClusterAddOnConditionAvailable,
+			Status:  metav1.ConditionTrue,
+			Reason:  addonapiv1alpha1.AddonAvailableReasonWorkApply,
+			Message: "Addon manifestWork is applied",
+		})
+		return nil
+	}
+
+	if !meta.IsStatusConditionTrue(addon.Status.Conditions, addonapiv1alpha1.ManagedClusterAddOnManifestApplied) {
```

Comment:
I think maybe we should only adding conditions of available when there is manifestapplied condition